Enable some imake builds on darwin

knedlsepp opened the pull request:
###### Motivation for this change
This enables a few imake based builds on darwin.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [x] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

LnL7 commented:
@GrahamcOfBorg build xskat

GrahamcOfBorg commented:
Failure on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.39173&attempt_id=c662db89-409b-4086-bb19-c3e663d83715)

Attempted: xskat

<details><summary>Partial log (click to expand)</summary><p>

```
make Makefiles
make: Nothing to be done for 'Makefiles'.
make includes
make: Nothing to be done for 'includes'.
make depend
gccmakedep  --   -I/nix/store/n8jlkihi2da4mfaj113xqvqc7qvnkgnb-xorg-cf-files-1.0.6/include    -Dlinux -D__aarch64__ -D_POSIX_C_SOURCE=199309L                           -D_POSIX_SOURCE -D_XOPEN_SOURCE                                 -D_BSD_SOURCE -D_SVID_SOURCE                                 -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64                                                                   -DFUNCPROTO=15 -DNARROWPROTO   -DDEFAULT_LANGUAGE=\"english\" -DDEFAULT_IRC_SERVER=\"irc.fu-berlin.de\"   -- skat.c ramsch.c null.c bitmaps.c xio.c xdial.c irc.c text.c
building
substitute(): ERROR: file 'makefile' does not exist
builder for '/nix/store/s6nw2kf93qqs94y76p9hf62nxiq6pmkq-xskat-4.0.drv' failed with exit code 1
[31;1merror:[0m build of '/nix/store/s6nw2kf93qqs94y76p9hf62nxiq6pmkq-xskat-4.0.drv' failed
```
</p></details>

GrahamcOfBorg commented:
Success on x86_64-darwin [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.39173&attempt_id=32f96e7c-6dd5-4cbe-a283-bbddfdb7f979)

Attempted: xskat

<details><summary>Partial log (click to expand)</summary><p>

```
+ mkdir -p /nix/store/8nkcz6ng7b6rgckl0jfd6darnm3nmrv1-xskat-4.0/man/man1
install -c -m 0444 xskat._man /nix/store/8nkcz6ng7b6rgckl0jfd6darnm3nmrv1-xskat-4.0/man/man1/xskat.1
install -c -m 0444 xskat-de._man /nix/store/8nkcz6ng7b6rgckl0jfd6darnm3nmrv1-xskat-4.0/man/man1/xskat-de.1
install.man in . done
post-installation fixup
moving /nix/store/8nkcz6ng7b6rgckl0jfd6darnm3nmrv1-xskat-4.0/man to /nix/store/8nkcz6ng7b6rgckl0jfd6darnm3nmrv1-xskat-4.0/share/man
gzipping man pages under /nix/store/8nkcz6ng7b6rgckl0jfd6darnm3nmrv1-xskat-4.0/share/man/
strip is /nix/store/kdff2gim6417493yha769kh00n63lnrw-cctools-binutils-darwin/bin/strip
stripping (with command strip and flags -S) in /nix/store/8nkcz6ng7b6rgckl0jfd6darnm3nmrv1-xskat-4.0/bin
patching script interpreter paths in /nix/store/8nkcz6ng7b6rgckl0jfd6darnm3nmrv1-xskat-4.0
```
</p></details>

LnL7 commented on pkgs/games/xskat/default.nix:
```diff
@@ -21,6 +21,10 @@ stdenv.mkDerivation {
     inherit (s) url sha256;
   };
+  preBuild = ''
+    substituteInPlace makefile --replace '/usr/bin/' "" \
```
`Makefile`?

knedlsepp replied:
Hm. The filename on darwin is actually lowercase "makefile". However replacing it with "Makefile" does work. Should I:
- Use "Makefile" and rely on the darwin filesystem being case-insensitive
- Make the preBuild phase optional stdenv.isDarwin?

LnL7 replied:
I only checked linux does it actually show up lowercase on darwin?

LnL7 replied:
`-rw-r--r--   1 lnl  staff    1679 Mar  8  2000 Makefile`

knedlsepp replied:
That's weird:
```
.....
-rw-r--r--  1 nixbld1 nixbld  29K Apr 19 21:52 makefile.bak
-rw-r--r--  1 nixbld1 nixbld  80K Apr 19 21:52 makefile
drwxr-xr-x 37 nixbld1 nixbld 1.3K Apr 19 21:52 .
```

knedlsepp replied:
When I put:
```
  preBuild = ''
    ls -rltah .
    exit 1
  '';
```

LnL7 replied:
preConfigure => `-rw-r--r--  1 nixbld1 nixbld   1679 Mar  8  2000 Makefile`
preBuild => `-rw-r--r--  1 nixbld1 nixbld  81101 Apr 20 00:12 makefile`

knedlsepp replied:
Ok. How about `substituteInPlace ?akefile`?

LnL7 replied:
I guess, but does it actually need to run after the configurePhase? otherwise `prePatch`/`postPatch` are more appropriate IMHO.

knedlsepp commented:
@LnL7 The more general solution to the problem: The `xorg.imake` derivation features a command `xmkmf`, which generates these makefiles from `xorgcffiles`. This one in turn contains a file `lib/X11/config/darwin.cf`, hard coding:
```
# define CppCmd                 /usr/bin/cpp3
# define CppCmd                 /usr/bin/cpp
#define CcCmd                   /usr/bin/cc
#define CplusplusCmd            /usr/bin/c++
```
So if we patch the darwin.cf file, we can probably make a lot more builds using `imake` succeed on darwin.

knedlsepp committed:
xorg.xorgcffiles: Don't use darwin's /usr/bin

All imake (xmkmf) based builds use the lib/X11/config/darwin.cf file to
define locations of cpp, cc, c++ (in /usr/bin by default). We remove the
directoy part to force darwin builds to search the $PATH for those
commands.

knedlsepp committed:
xskat: enable darwin build

knedlsepp committed:
oneko: enable darwin build

knedlsepp committed:
Xaw3d: enable darwin build

knedlsepp committed:
gv: enable darwin build

knedlsepp committed:
xsok: enable darwin build

knedlsepp commented:
Well. I think that's mostly it. The other programs don't seem overly interesting.

LnL7 commented:
@GrahamcOfBorg build xorg.xorgcffiles xskat oneko Xaw3d gv xsok

GrahamcOfBorg commented:
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.39173&attempt_id=34b2afd3-27a9-40ba-b8e5-68a32bcd8cdf)

Attempted: xorg.xorgcffiles, xskat, oneko, Xaw3d, gv, xsok

<details><summary>Partial log (click to expand)</summary><p>

```
stripping (with command strip and flags -S) in /nix/store/jkhj630dc2kfqyafq931m38ma8pq1y07-gv-3.7.4/bin
patching script interpreter paths in /nix/store/jkhj630dc2kfqyafq931m38ma8pq1y07-gv-3.7.4
/nix/store/k41w8cv366mfqpys7xh7f79kh3sw18qa-xorg-cf-files-1.0.6
/nix/store/jkhj630dc2kfqyafq931m38ma8pq1y07-gv-3.7.4/bin/gv-update-userconfig: interpreter directive changed from " /usr/bin/perl" to "/nix/store/cxdmh98g0lvl1dyq304c1lq7f90dh01f-perl-5.24.3/bin/perl"
/nix/store/xpn78hkdslbwlzypyzvahkfxgwpmzh7d-xskat-4.0
checking for references to /build in /nix/store/jkhj630dc2kfqyafq931m38ma8pq1y07-gv-3.7.4...
/nix/store/4sks2689nv5l1n0b3snga8jmfy56d3m3-oneko-1.2.5
/nix/store/cgfc4m6grlqmxnrp5x9l7scy6ivlj0i7-Xaw3d-1.6.2
/nix/store/jkhj630dc2kfqyafq931m38ma8pq1y07-gv-3.7.4
/nix/store/8wrlr3fkxvcvlzrpx0gkljpgqw4qq9qp-xsok-1.02
```
</p></details>

GrahamcOfBorg commented:
Success on x86_64-darwin [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.39173&attempt_id=e66264d3-2f2e-42c3-98b7-3e138593baf3)

Attempted: xorg.xorgcffiles, xskat, oneko, Xaw3d, gv, xsok

<details><summary>Partial log (click to expand)</summary><p>

```
strip is /nix/store/kdff2gim6417493yha769kh00n63lnrw-cctools-binutils-darwin/bin/strip
stripping (with command strip and flags -S) in /nix/store/8406ry2rv0pzbv0840v7khmglwz3lrrc-gv-3.7.4/bin
patching script interpreter paths in /nix/store/8406ry2rv0pzbv0840v7khmglwz3lrrc-gv-3.7.4
/nix/store/8406ry2rv0pzbv0840v7khmglwz3lrrc-gv-3.7.4/bin/gv-update-userconfig: interpreter directive changed from " /usr/bin/perl" to "/nix/store/lizm7q36j5sf5524bw9gsykflv9i2l7z-perl-5.24.3/bin/perl"
/nix/store/0h83p7wskip2d39qzsfz4ddpn7i0yjx6-xorg-cf-files-1.0.6
/nix/store/zsd9jlhnj31affm2bzf0pkhjfdn75vnh-xskat-4.0
/nix/store/3pl3ksn76fvrwlfqcsf0cma7izg17s1k-oneko-1.2.5
/nix/store/hlv6ffrfyb4a5iaaz1g640k8srq430kp-Xaw3d-1.6.2
/nix/store/8406ry2rv0pzbv0840v7khmglwz3lrrc-gv-3.7.4
/nix/store/rz4brxmi8ks2qym7ah1bpvgryk2s3qsr-xsok-1.02
```
</p></details>

GrahamcOfBorg commented:
Success on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.39173&attempt_id=50ce6d35-8f35-4b42-9734-1e3578e235ba)

Attempted: xorg.xorgcffiles, xskat, oneko, Xaw3d, gv, xsok

<details><summary>Partial log (click to expand)</summary><p>

```
stripping (with command strip and flags -S) in /nix/store/7gdrgsh111haj9aiavpca8n0r8h7vfsz-gv-3.7.4/bin
patching script interpreter paths in /nix/store/7gdrgsh111haj9aiavpca8n0r8h7vfsz-gv-3.7.4
/nix/store/7gdrgsh111haj9aiavpca8n0r8h7vfsz-gv-3.7.4/bin/gv-update-userconfig: interpreter directive changed from " /usr/bin/perl" to "/nix/store/3qhzffjvl8gbiabikiqdnz4jhh9v84nz-perl-5.24.3/bin/perl"
checking for references to /build in /nix/store/7gdrgsh111haj9aiavpca8n0r8h7vfsz-gv-3.7.4...
/nix/store/djn1mp567yh5mih7mr8sczk1pnr0di4b-xorg-cf-files-1.0.6
/nix/store/hk64bsg1lxlfrd68133kljbmg8w5k4pc-xskat-4.0
/nix/store/q107w5yb39n3cpryh31jlcw0qhz9n4m9-oneko-1.2.5
/nix/store/lnyj8qd3xq796d6y76d47xvazz1vqh61-Xaw3d-1.6.2
/nix/store/7gdrgsh111haj9aiavpca8n0r8h7vfsz-gv-3.7.4
/nix/store/3v1z89x6rb9hh1bch3pmkrabi0b526nd-xsok-1.02
```
</p></details>
